src/category_theory/limits/shapes/constructions/limits_of_products_and_equalizers.lean

```diff
@@ -0,0 +1,92 @@
+/-
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison
+-/
+
+import category_theory.limits.shapes.products
+import category_theory.limits.shapes.equalizers
+
+/-!
+# Constructing limits from products and equalizers.
+
+If a category has all products, and all equalizers, then it has all limits.
+
+TODO: provide the dual result.
+-/
+
+open category_theory
+open opposite
+
+namespace category_theory.limits
+
+universes v u
+variables {C : Type u} [𝒞 : category.{v+1} C]
+include 𝒞
+
+@[simp] def equalizer_diagram [has_products.{v} C] {J} [small_category J] (F : J ⥤ C) : walking_parallel_pair ⥤ C :=
+let pi_obj := limits.pi_obj F.obj in
+let pi_hom := limits.pi_obj (λ f : (Σ p : J × J, p.1 ⟶ p.2), F.obj f.1.2) in
+let s : pi_obj ⟶ pi_hom :=
+  pi.lift (λ f : (Σ p : J × J, p.1 ⟶ p.2), pi.π F.obj f.1.1 ≫ F.map f.2) in
+let t : pi_obj ⟶ pi_hom :=
+  pi.lift (λ f : (Σ p : J × J, p.1 ⟶ p.2), pi.π F.obj f.1.2) in
+parallel_pair s t
+
+@[simp] def equalizer_diagram.cones_hom [has_products.{v} C] {J} [small_category J] (F : J ⥤ C) :
+  (equalizer_diagram F).cones ⟶ F.cones :=
+{ app := λ X c,
+  { app := λ j, c.app walking_parallel_pair.zero ≫ pi.π _ j,
+    naturality' := λ j j' f,
+    begin
+      have L := c.naturality walking_parallel_pair_hom.left,
+      have R := c.naturality walking_parallel_pair_hom.right,
+      have t := congr_arg (λ g, g ≫ pi.π _ (⟨(j, j'), f⟩ : Σ (p : J × J), p.fst ⟶ p.snd)) (R.symm.trans L),
+      dsimp at t,
```

Comment:
Can the proof be changed into a `simpa only [foobars] using t` from here?

Reply:
Needed an extra `dsimp`, but yes. I guess `simpa` is outside my usual toolkit; seems like abbreviation for abbreviation's sake.